example/pub-sub-chat.ss

```
@@ -0,0 +1,143 @@
+
+  (try
+   (let ((values cmd opt) (getopt-parse gopt args))
+     (let-hash opt
+       (case cmd
+         ((bootstrap) ;;In case where the command is to bootstrap
+          (do-startup #f .host-addresses))
+         ((connect) ;;where command is to connect
+          (do-startup .peer .host-addresses))
+         ((help) ;;where command is to help
+          (getopt-display-help-topic gopt .?command "pub-sub-chat")))))
+   (catch (getopt-error? exn)
+     (getopt-display-help exn "pub-sub-chat" (current-error-port))
+     (exit 1))))
+
+
+;; do-startup
+;; Parameters:
+;; peer: #f or a peer Addresss
+;; host-addresses: a list of multiaddresses
+;;
+;;If peer is #f, do-startup will not try to connect to any peers and will subscribe to the topic "chat"
+;;
+;;If peer is anything else, do-startup will try to connect to peer and will subscribe to the topic "chat"
+(def (do-startup peer host-addresses)
+   (displayln "starting up")
+
+   ;start the daemon with custom arguments -pubsub and -connManager to enable pubsub
+   (start-libp2p-daemon! host-addresses: host-addresses options: ["-pubsub" "-connManager"] wait: 10)
+
+   ;open a client with the given host-address, it will use the daemon already running
+  (let* ((c (open-libp2p-client host-addresses: host-addresses wait: 10))
+         (self (libp2p-identify c)))
+    (for (p (peer-info->string* self))
+      (displayln "I am " p)) ;Identify yourself. Print each multiaddr
+
+    (if peer ;If there is a bootstrap peer (connecting command) then connect to it
+      (begin
+        (displayln "Bootsrap peer found, connecting to it")
+        (libp2p-connect c peer)
+        (displayln "Connected to Bootstrap peer")))
+
+    ;Subscribe to the chat and spawn a reader thread and writer thread
+    (let*-values (((sub cancel) (pubsub-subscribe c "chat"))
+           (reader (spawn subscription-reader sub self)))
+      (subscription-writer c)
+      (thread-terminate! reader)))) ;;TODO: Does this exit cleanly?
```

no, it really abruptly kills the thread -- you might want to raise an exception in it to interrupt it. See `:std/misc/threads`, you can use `thread-raise!` or `thread-abort!`.

---

Sounds good! Is there a way to make sure the daemon exits cleanly as well? Or will this do it?

---

Just pushed your suggestions!

---

i think it should be fine, but might make sense to explicitly close the client somewhere.